File: Day 1/part2.ts
// const input = `two1nine
// eightwothree
// abcone2threexyz
// xtwone3four
// 4nineeightseven2
// zoneight234
// 7pqrstsixteen`.split("\n");

const input = await Deno.readTextFile("input.txt").then((text) =>
  text.split("\r\n")
);

// console.log(input);

const calibrationValues: number[] = [];

function findDigits(s: string, digits: string[] = []) {
  if (!s || s.length === 0) return digits;
  if (!isNaN(+s[0])) {
    digits.push(s[0]);
    return findDigits(s.slice(1), digits);
  } else if (s.startsWith("one")) {
    digits.push("1");
    return findDigits(s.slice(1), digits);
  } else if (s.startsWith("two")) {
    digits.push("2");
    return findDigits(s.slice(1), digits);
  } else if (s.startsWith("three")) {
    digits.push("3");
    return findDigits(s.slice(1), digits);
  } else if (s.startsWith("four")) {
    digits.push("4");
    return findDigits(s.slice(1), digits);
  } else if (s.startsWith("five")) {
    digits.push("5");
    return findDigits(s.slice(1), digits);
  } else if (s.startsWith("six")) {
    digits.push("6");
    return findDigits(s.slice(1), digits);
  } else if (s.startsWith("seven")) {
    digits.push("7");
    return findDigits(s.slice(1), digits);
  } else if (s.startsWith("eight")) {
    digits.push("8");
    return findDigits(s.slice(1), digits);
  } else if (s.startsWith("nine")) {
    digits.push("9");
    return findDigits(s.slice(1), digits);
  } else {
    return findDigits(s.slice(1), digits);
  }
}

for (let i = 0; i < input.length; i++) {
  const digits: string[] = findDigits(input[i]);

  const first = digits[0];
  const second = digits[digits.length - 1];

  const calibrationValue = first + second;
  calibrationValues.push(Number(calibrationValue));
}

const sum = calibrationValues.reduce((a, b) => a + b);
console.log(sum);

File: Day 1/part2.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
	"unicode"
)

func Part2() {
	content, err := os.ReadFile("input.txt")

	if err != nil {
		log.Fatal(err)
	}

	lines := strings.Split(string(content), "\n")

	var sum uint = 0

	for _, line := range lines {
		var firstDigit uint = 00
		var lastDigit uint = 00

		for idx, char := range line {
			if unicode.IsDigit(char) && firstDigit == 00 {
				digit, _ := strconv.Atoi(string(char))
				firstDigit = uint(digit)
			}

			if unicode.IsDigit(char) {
				digit, _ := strconv.Atoi(string(char))
				lastDigit = uint(digit)
			}

			// Check for spelled digits
			s := line[idx:]

			if strings.HasPrefix(s, "one") {
				if firstDigit == 00 {
					firstDigit = 1
				}
				lastDigit = 1
				continue
			}

			if strings.HasPrefix(s, "two") {
				if firstDigit == 00 {
					firstDigit = 2
				}
				lastDigit = 2
				continue
			}

			if strings.HasPrefix(s, "three") {
				if firstDigit == 00 {
					firstDigit = 3
				}
				lastDigit = 3
				continue
			}

			if strings.HasPrefix(s, "four") {
				if firstDigit == 00 {
					firstDigit = 4
				}
				lastDigit = 4
				continue
			}

			if strings.HasPrefix(s, "five") {
				if firstDigit == 00 {
					firstDigit = 5
				}
				lastDigit = 5
				continue
			}

			if strings.HasPrefix(s, "six") {
				if firstDigit == 00 {
					firstDigit = 6
				}
				lastDigit = 6
				continue
			}

			if strings.HasPrefix(s, "seven") {
				if firstDigit == 00 {
					firstDigit = 7
				}
				lastDigit = 7
				continue
			}

			if strings.HasPrefix(s, "eight") {
				if firstDigit == 00 {
					firstDigit = 8
				}
				lastDigit = 8
				continue
			}

			if strings.HasPrefix(s, "nine") {
				if firstDigit == 00 {
					firstDigit = 9
				}
				lastDigit = 9
				continue
			}
		}

		sum += firstDigit*10 + lastDigit
	}

	fmt.Printf("Part 2 > Sum of calibration values: %v\n", sum)
}
